var/spack/repos/builtin/packages/mfem/package.py

```diff
@@ -267,16 +267,12 @@ class Mfem(Package, CudaPackage, ROCmPackage):
     # Propagate 'cuda_arch' to 'hypre' without propagating the '+cuda'
     # variant because we want to allow 'mfem+cuda ^hypre~cuda':
     for sm_ in CudaPackage.cuda_arch_values:
-        depends_on(
-            "hypre@2.22.1:+cuda cuda_arch={0}".format(sm_),
-            when="+mpi+cuda cuda_arch={0} ^hypre+cuda".format(sm_),
-        )
+        depends_on(f"hypre@2.22.1:+cuda cuda_arch={sm_}", when=f"+mpi+cuda cuda_arch={sm_}")
```

Comment:
These edits change the desired behavior. We want:
1. `mfem+cuda cuda_arch=xx` to use `hypre` without forcing `hypre+cuda`, i.e. build `mfem` with cuda but `hypre` without cuda
2. `mfem+cuda cuda_arch=xx ^hypre+cuda` to propagate the `cuda_arch=xx` flag to `hypre`.

Reply:
Ok, that wasn't clear from the directive, sorry (in particular, enforcing `+cuda` when `when=+cuda` is tautological). 

What about the other constraints that are enforced e.g. `@2.22.1:`? Currently they are enforced only if `hypre` is using `+cuda`. Is that correct?

Reply:
> Ok, that wasn't clear from the directive, sorry (in particular, enforcing `+cuda` when `when=+cuda` is tautological).

It is not tautological: one `+cuda` applies to `hypre`, the other `+cuda` (in the `when` option) applies to `mfem`. The comment just above the loop is trying describe what's happening there but it may be not that clear.

> What about the other constraints that are enforced e.g. `@2.22.1:`? Currently they are enforced only if `hypre` is using `+cuda`. Is that correct?

Yes, to use `hypre+cuda`, `mfem` needs `hypre` version as indicated.